lib/vsc/filesystem/oceanstor.py

```diff
@@ -1810,3 +1810,141 @@ def _get_grace_expiration(grace, usage, soft, hard):
             expired = (False, None)
 
         return expired
+
+    def list_snapshots(self, filesystem, fileset=None):
+        """
+        List the snapshots in the given filesystem or dtree fileset
+
+        @type filesystem: name of the filesystem
+        @type fileset: name of the dtree fileset
+        """
+
+        fs = self.get_filesystem_info(filesystem)
+        filter_json = {"file_system_id": int(fs["id"])}
+
+        if fileset is not None:
+            dtree = self.get_fileset_info(filesystem, fileset)
+            if dtree is None:
+                err_msg = "Snapshot query failed: fileset '%s' not found in filesystem '%s'" % (filesystem, fileset)
+                self.log.raiseException(err_msg, OceanStorOperationError)
+            filter_json.update({"dtree_id": dtree["id"]})
+
+        filter_json = json.dumps([filter_json], separators=OCEANSTOR_JSON_SEP)
+        _, response = self.session.file_service.snapshots.get(pagination=True, filter=filter_json)
+
+        snapshots = [snap["name"] for snap in response["data"]]
+
+        return snapshots
+
+    def create_filesystem_snapshot(self, fsname, snapname, filesets=None):
+        """
+        Create a filesystem snapshot. If filesets is None, it's full system snapshots
+        else the snapshot is limited to the list of filesets given.
+
+        @type fsname: string representing the name of the filesystem
+        @type snapname: string representing the name of the new snapshot
+        @type filesets: list of fileset names to take a snapshots of
+        """
+        if filesets is not None:
+            # fileset/dtree snapshot
+            if not isinstance(filesets, list):
+                filesets = [filesets]
+
+            for fileset in filesets:
+                if self.get_fileset_info(fsname, fileset) is None:
+                    self.log.error("Cannot create snapshot: fileset %s not found on filesystem %s!", fileset, fsname)
+                    return 0
+
+                # the snapshot namespace of all filesets is shared in OceanStor
+                # create unique snapshot name per fileset
+                fileset_snapname = "%s_%s" % (fileset, snapname)
```

Comment:
I would definitely add debug print that you're doing this?

Reply:
good point, I also made a new private method to make this name change. Fixed in https://github.com/vub-hpc/vsc-filesystem-oceanstor/pull/24/commits/3a01e6dd3325f8252670d28c9f21a389a89f06b8